piecrust-uplink/src/abi/state.rs

```diff
@@ -179,6 +210,33 @@ pub fn call_raw(
     call_raw_with_limit(contract, fn_name, fn_arg, 0)
 }
 
+/// Same as [`call_raw`] but in addition to a called
+/// function return value, it also returns the contents
+/// of buffer B
+pub fn call_raw_b(
+    contract: ContractId,
+    fn_name: &str,
+    fn_arg: &[u8],
+) -> Result<(Vec<u8>, u64, u64), ContractError> {
+    call_raw_with_limit_b(contract, fn_name, fn_arg, 0)
+}
+
+/// Allows the contract to set allowance which will be used
+/// to pay for the current call, under the condition that contract's
+/// current balance is greater or equal to the allowance
+/// and the allowance is sufficient to cover gas cost.
+/// This call is of no consequence if the above conditions are not met.
+pub fn set_allowance(b: u64) {
+    set_b(b, BIndex::Allowance);
+}
+
+/// Allows the contract to set charge which will be used
+/// as a fee. If charge is greater than gas spent,
+/// the difference will be added to contract's balance.
+pub fn set_charge(b: u64) {
+    set_b(b, BIndex::Charge);
+}
```

Comment:
I decided to leave these functions as separate ones, because I think they are more contract writer-friendly this way IMO. Let's give it a second look.